src/parse.rs

```diff
@@ -99,116 +99,118 @@ impl AST {
-                            continue;
-                        }
-                    };
-                    language = lang;
-
-                    for component in components.into_iter() {
-                        match component {
-                            ComponentType::ClassOrInterfaceComponent(component) => {
-                                match component.declaration_type {
-                                    ContainerType::Class => {
-                                        module.classes.push(component);
-                                    }
-                                    ContainerType::Interface => {
-                                        module.interfaces.push(component);
-                                    }
-                                    r#type => {
-                                        println!("got other label when it should have been class/ifc: {:#?}", r#type);
-                                    }
+        // Generate module identifier
+        let p = dir.path.clone();
+        let mod_path;
+        if p.is_file() {
+            let p: PathBuf = p.into_iter().dropping_back(1).collect();
+            mod_path = p.into_os_string().into_string().unwrap()
+        } else {
+            mod_path = path.clone();
+        }
+        let module_name = mod_path.clone();
+
+        // Get directory
+        let read_dir = std::fs::read_dir(dir.path)?;
+        let mut module = ModuleComponent::new(module_name.to_string(), path);
+
+        for entry in read_dir {
+            let entry = entry?;
+            if !entry.path().is_dir() {
+                if dir
+                    .files
+                    .iter()
+                    .find(|path_buf| path_buf == &&entry.path())
+                    .is_none()
+                {
+                    continue;
+                }
+                let mut file = File::open(entry.path())?;
+                let (components, lang) = match parse_file(&mut file, &entry.path()) {
+                    Ok(res) => res,
+                    Err(err) => {
+                        eprintln!("Could not read file {:?}: {:#?}", entry.path(), err);
+                        continue;
+                    }
+                };
+                language = lang;
+
+                for component in components.into_iter() {
+                    match component {
+                        ComponentType::ClassOrInterfaceComponent(component) => {
+                            match component.declaration_type {
+                                ContainerType::Class => {
+                                    module.classes.push(component);
+                                }
+                                ContainerType::Interface => {
+                                    module.interfaces.push(component);
+                                }
+                                r#type => {
+                                    println!(
```

Comment:
This reports an error if you don't have a Class or Interface component. However, in Java, I'd expect to see 'AnnotationComponent's (since an Annotation is a special class), unless we decide to just portray Annotations as classes and reserve AnnotationComponent for an annotation on an object/field/method. Does one solution sound better than the other to you?

Reply:
I don't think this was handled in the original JParser -- so I would just model it as a class. It would be good to ask Vincent about this, though.

Reply:
So according to Dr. Cerny in today's standup meeting, annotation definitions are currently ignored. We could ignore them for now, but for future extension we may want to discuss either modeling them just as a class or adding a new structure.